src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,135 @@
 namespace chip {
 namespace app {
 
-class AttributeValueEncoder : protected TagBoundEncoder
+/**
+ * The AttributeVValueEncoder is a helper class for filling report payloads into AttributeReportIBs.
+ * The attribute value encoder can be initialized with a AttributeEncodeState for saving and recovering its state between encode
+ * sessions (chunkings).
+ *
+ * When Encode returned recoverable errors (e.g. CHIP_ERROR_NO_MEMORY) the state can be used to initialize the AttributeValueEncoder
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
 {
 public:
-    AttributeValueEncoder(TLV::TLVWriter * aWriter, FabricIndex aAccessingFabricIndex) :
-        TagBoundEncoder(aWriter, TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData))),
-        mAccessingFabricIndex(aAccessingFabricIndex)
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
+    };
+
+    AttributeValueEncoder(AttributeReportIBs::Builder & aAttributeReportIBsBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, DataVersion aDataVersion,
+                          const AttributeEncodeState & aState = AttributeEncodeState()) :
+        mAttributeReportIBsBuilder(aAttributeReportIBsBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mDataVersion(aDataVersion), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+
+        AttributeReportIB::Builder attributeReportIBBuilder = mAttributeReportIBsBuilder.CreateAttributeReport();
```

Comment:
This seems like a _lot_ of code to inline in every caller of this method.  And most of it does not depend on `Ts` in any way, right?  It seems like we should refactor this to have a non-template superclass that does a lot of this work and a templated subclass that does as little as it can.

That would also let us out-of-line the code, which would help with readability in this file, which is starting to suffer.

Or maybe this state management should stay in ReadSingleClusterData in shared code?  The hard part is that there we don't know whether the callee will encode the whole list or individual items?

Reply:
Made some changes to the code to offload some code to separate functions, this should make the Encode a bit clearer.

Reply:
Created `AttributeReportBuilder` for actual encoding. Now AttributeValueEncoder::Encode only has 4 lines, should be clearer.